Construindo a Rede Neura
========================

Redes neurais são compostas por camadas/módulos que realizam operações em dados. 
O namespace [torch.nn](https://pytorch.org/docs/stable/nn.html) fornece todos 
os blocos de construção necessários para construir sua própria rede neural. 
Cada módulo no PyTorch é uma subclasse de
[nn.Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html).
Essa estrutura aninhada permite construir e gerenciar arquiteturas complexas facilmente.

Nas seções a seguir, construiremos uma rede neural para classificar imagens no conjunto de dados FashionMNIST.

In [17]:
import os
import torch
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

Obtenha o dispositivo para treinamento
=======================

Queremos poder treinar nosso modelo em um
[acelerador](https://pytorch.org/docs/stable/torch.html#accelerators)
como CUDA, MPS, MTIA ou XPU. Se o acelerador atual estiver disponível, 
o usaremos. Caso contrário, usaremos a CPU.


In [18]:
device = torch.accelerator.current_accelerator().type if torch.accelerator.is_available() else "cpu" # type: ignore
#device = "cuda" if torch.accelerator.is_available() else "cpu"
print(f"Using {device} device")

Using cuda device


Defina a Classe
================

Definimos nossa rede neural subclassificando  `nn.Module`, 
e inicializando as camadas da rede neural em `__init__`. Cada subclass `nn.Module`
implementa as operações nos dados de entrada do método `forward`.

In [19]:
class NeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28*28, 512),
            nn.ReLU(),
            nn.Linear(512, 512),
            nn.ReLU(),
            nn.Linear(512, 10),
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits

Criamos uma instância de `NeuralNetwork`, movemos para o `device`,
e imprimimos sua estrutura.

In [20]:
model = NeuralNetwork().to(device)
print(model)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (0): Linear(in_features=784, out_features=512, bias=True)
    (1): ReLU()
    (2): Linear(in_features=512, out_features=512, bias=True)
    (3): ReLU()
    (4): Linear(in_features=512, out_features=10, bias=True)
  )
)


Para usar o modelo, passamos os dados de entrada. Isso executa o
`forward`, juntamente com algumas 
[operações em segundo plano ](https://github.com/pytorch/pytorch/blob/270111b7b611d174967ed204776985cefca9c144/torch/nn/modules/module.py#L866).
Não chame `model.forward()` diretamente!

A chamada do modelo na entrada retorna um tensor bidimensional 
com dim=0 correspondendo a cada saída de 10 valores brutos 
previstos para cada classe, e dim=1 correspondendo aos valores 
individuais de cada saída. Obtemos as probabilidades de previsão 
passando-as por uma instância do módulo `nn.Softmax`.

In [21]:
X = torch.rand(1, 28, 28, device=device)
logits = model(X)
pred_probab = nn.Softmax(dim=1)(logits)
y_pred = pred_probab.argmax(1)
print(f"Predicted class: {y_pred}")

Predicted class: tensor([5], device='cuda:0')


------------------------------------------------------------------------


Camadas do modelo
============

Vamos analisar as camadas do modelo FashionMNIST. Para ilustrar, pegaremos um minilote de amostra com 3 imagens de tamanho 28x28 e veremos o que acontece com ele à medida que o passamos pela rede.

In [22]:
input_image = torch.rand(3,28,28)
print(input_image.size())

torch.Size([3, 28, 28])


nn.Flatten
==========

Inicializamos a camada 
[nn.Flatten](https://pytorch.org/docs/stable/generated/torch.nn.Flatten.html)
para converter cada imagem 2D 28x28 em uma matriz contígua de 784 
valores de pixels (a dimensão do minibatch (em dim=0) é mantida).

In [23]:
flatten = nn.Flatten()
flat_image = flatten(input_image)
print(flat_image.size())

torch.Size([3, 784])


nn.Linear
=========

A [Camada linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html)
é um módulo que aplica uma transformação linear na entrada usando seus pesos e vieses armazenados.

In [24]:
layer1 = nn.Linear(in_features=28*28, out_features=20)
hidden1 = layer1(flat_image)
print(hidden1.size())

torch.Size([3, 20])


nn.ReLU
=======

Ativações não lineares são o que criam os mapeamentos complexos 
entre as entradas e saídas do modelo. Elas são aplicadas após 
transformações lineares para introduzir a **não linearidade**, 
ajudando as redes neurais a aprender uma ampla variedade de fenômenos.

Neste modelo, usamos
[nn.ReLU](https://pytorch.org/docs/stable/generated/torch.nn.ReLU.html)
entre nossas camadas lineares, mas há outras ativações 
para introduzir não linearidade em seu modelo.

In [25]:
print(f"Before ReLU: {hidden1}\n\n")
hidden1 = nn.ReLU()(hidden1)
print(f"After ReLU: {hidden1}")

Before ReLU: tensor([[-1.4978e-01,  3.5953e-01, -2.5358e-01,  2.2685e-01,  2.1614e-01,
          6.2577e-01,  1.0041e-01,  2.3143e-02,  4.7045e-01,  1.7473e-01,
         -2.3214e-01, -1.6198e-01,  2.5856e-01,  1.4889e-01, -2.8388e-01,
         -2.8260e-02,  5.2430e-02,  1.0162e-01, -3.2427e-01, -2.7101e-01],
        [ 1.4328e-01,  1.1175e+00,  1.0312e-01,  1.4313e-01,  2.4551e-03,
          6.0019e-01,  6.5454e-02, -2.4490e-02,  4.1264e-01, -8.2317e-02,
         -2.7198e-01, -3.4181e-01,  2.4958e-01,  5.3462e-01,  7.4921e-03,
         -1.0417e-03, -9.8114e-02,  3.6218e-01, -3.8502e-01,  1.0535e-01],
        [-3.7121e-01,  6.5653e-01,  6.7728e-02, -1.6095e-01,  3.0306e-02,
          5.7243e-01,  4.3165e-01, -3.6883e-01,  1.5029e-01,  1.8605e-01,
         -1.8206e-01,  2.6177e-01,  3.8354e-01,  3.7296e-01,  2.1004e-02,
         -2.5384e-01,  2.0413e-01,  3.5992e-01, -5.6264e-01,  1.1953e-01]],
       grad_fn=<AddmmBackward0>)


After ReLU: tensor([[0.0000, 0.3595, 0.0000, 0.2269, 0.2161,

nn.Sequential
=============

[nn.Sequential](https://pytorch.org/docs/stable/generated/torch.nn.Sequential.html)
 é um contêiner ordenado de módulos. Os dados são passados ​​por todos os módulos na mesma ordem definida. Você pode usar contêineres sequenciais para montar uma rede rápida como `seq_modules`.


In [26]:
seq_modules = nn.Sequential(
    flatten,
    layer1,
    nn.ReLU(),
    nn.Linear(20, 10)
)
input_image = torch.rand(3,28,28)
logits = seq_modules(input_image)

nn.Softmax
==========

A última camada linear da rede neural retorna `logits` valores brutos em
 \[-infty, infty\] que são passados ​​para o módulo
[nn.Softmax](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html).
Os logits são escalados para valores [0, 1] que representam as probabilidades previstas do modelo para cada classe. O parâmetro `dim`, indica a dimensão ao longo da qual os valores devem somar 1.


In [27]:
softmax = nn.Softmax(dim=1)
pred_probab = softmax(logits)

Parâmetros do modelo
================

Muitas camadas dentro de uma rede neural são parametrizadas, ou seja, 
possuem pesos e vieses associados que são otimizados durante o treinamento.
A subclassificação `nn.Module` rastreia automaticamente todos os campos 
definidos dentro do objeto do seu modelo e torna todos os parâmetros 
acessíveis usando os método `parameters()` ou `named_parameters()`.

Neste exemplo, iteramos sobre cada parâmetro e imprimimos seu tamanho e uma prévia de seus valores.

In [28]:
print(f"Model structure: {model}\n\n")

for name, param in model.named_parameters():
    print(f"Layer: {name} | Size: {param.size()} | Values : {param[:2]} \n")

Model structure: NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (0): Linear(in_features=784, out_features=512, bias=True)
    (1): ReLU()
    (2): Linear(in_features=512, out_features=512, bias=True)
    (3): ReLU()
    (4): Linear(in_features=512, out_features=10, bias=True)
  )
)


Layer: linear_relu_stack.0.weight | Size: torch.Size([512, 784]) | Values : tensor([[-0.0064, -0.0231,  0.0125,  ...,  0.0127, -0.0256, -0.0146],
        [ 0.0229,  0.0135,  0.0156,  ...,  0.0313, -0.0273,  0.0254]],
       device='cuda:0', grad_fn=<SliceBackward0>) 

Layer: linear_relu_stack.0.bias | Size: torch.Size([512]) | Values : tensor([ 0.0199, -0.0352], device='cuda:0', grad_fn=<SliceBackward0>) 

Layer: linear_relu_stack.2.weight | Size: torch.Size([512, 512]) | Values : tensor([[-0.0035,  0.0376, -0.0332,  ...,  0.0036, -0.0296, -0.0129],
        [-0.0221,  0.0148,  0.0388,  ..., -0.0241, -0.0022, -0.0125]],
       device='cuda:0', grad_fn=<Sl

------------------------------------------------------------------------


Leitura adicional
===============

-   [torch.nn API](https://pytorch.org/docs/stable/nn.html)
